cli/internal/run/real_run.go

```diff
@@ -210,12 +215,26 @@ func (ec *execContext) exec(ctx gocontext.Context, packageTask *nodes.PackageTas
 		ErrorPrefix:  prettyPrefix,
 		WarnPrefix:   prettyPrefix,
 	}
-	hit, err := taskCache.RestoreOutputs(ctx, prefixedUI, progressLogger)
-	if err != nil {
-		prefixedUI.Error(fmt.Sprintf("error fetching from cache: %s", err))
-	} else if hit {
-		tracer(TargetCached, nil)
-		return nil
+
+	cacheHit := taskCache.IsCacheHit(ctx)
+	outputtedCacheStatusLogs := false
+
+	// If we want grouped log ordering we want to either:
+	// - restore
+	// - not display any logs and do this when we have finished execution if we have a cache mis all output and display logs if we have a cache hit.s.
+	if cacheHit || !grouped {
+		outputtedCacheStatusLogs = true
+		logMutex.Lock()
+		hit, err := taskCache.RestoreOutputs(ctx, prefixedUI, progressLogger, false)
```

Comment:
I don't think we should be holding a lock while doing file IO (`RestoreOutputs`). This is going to serialize the file IO for all of the cache hits.

Reply:
You're right, this is not a good idea. It was the only way to do this when using mutexes, the problem will be solved when using the queue based method.